dali/base/dasess.cpp

```diff
@@ -1267,9 +1311,246 @@ class CLdapWorkItem : public Thread
+                        break;
+                    case DaliServerProcess:
+                        addRoles(component, { DCR_DaliServer, DCR_DaliDiag, DCR_SwapNode, DCR_UpdateEnv, DCR_DaliAdmin, DCR_TreeView, DCR_Testing, DCR_DaFsControl, DCR_XRef, DCR_Config, DCR_ScheduleAdmin });
+                        break;
+                }
+            }
+        }
+        IPropertyTree *whiteListTree = conn->queryRoot()->queryPropTree("WhiteList");
+        if (whiteListTree)
+        {
+            enabled = whiteListTree->getPropBool("@enabled", true); // on by default
+            Owned<IPropertyTreeIterator> whiteListIter = whiteListTree->getElements("Entry");
+            ForEach(*whiteListIter)
+            {
+                const IPropertyTree &entry = whiteListIter->query();
+                StringArray hosts, roles;
+                hosts.appendListUniq(entry.queryProp("@hosts"), ",");
+                roles.appendListUniq(entry.queryProp("@roles"), ",");
+                ForEachItemIn(h, hosts)
+                {
+                    ForEachItemIn(r, roles)
+                    {
+                        const char *roleStr = roles.item(r);
+                        const auto &it = daliClientRoleMap.find(roleStr);
+                        if (it != daliClientRoleMap.end())
+                        {
+                            IpAddress ip(hosts.item(h));
+                            if (!ip.isNull())
+                            {
+                                StringBuffer ipStr;
+                                whiteList.insert({ ip.getIpText(ipStr).str(), it->second });
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        populated = true;
+    }
+    void ensurePopulated() const
+    {
+        // should be called within CS
+        if (populated)
+            return;
+        (const_cast <CWhiteListHandler *> (this))->populate();
+    }
+public:
+    bool isWhiteListed(const char *ip, DaliClientRole role) const
+    {
+        CriticalBlock block(populatedCrit);
+        ensurePopulated();
+        const auto &it = whiteList.find({ip, role});
+        if (it != whiteList.end())
+            return true;
+        else if (enabled)
+            return false;
+        else
+        {
+            WARNLOG("Whitelist mechanism is currently disabled");
+            return true;
+        }
+    }
+    StringBuffer &getWhiteList(StringBuffer &out) const
+    {
+        CriticalBlock block(populatedCrit);
+        ensurePopulated();
+        for (auto &it: whiteList)
+            out.append(it.first.c_str()).append(", ").append(queryRoleName(it.second)).append("\n");
+        return out;
+    }
+    void refresh()
+    {
+        /* NB: clear only, so that next usage will re-populated
+         * Do not want to repopulate now, because refresh() is likely called within a update write transaction
+         */
+        CriticalBlock block(populatedCrit);
```

Comment:
How common is it likely to be that refresh will block the whitelisting checks?  If common you could use a different approach to reduce the scope of the critical section.

Reply:
I do not think it will be refreshing often, or that client registration time will be so rapid and frequent that it will be an issue.